src/merkle/index.rs

```diff
@@ -18,20 +29,30 @@ impl NodeIndex {
     // --------------------------------------------------------------------------------------------
 
     /// Creates a new node index.
-    pub const fn new(depth: u8, value: u64) -> Self {
-        Self { depth, value }
+    ///
+    /// # Errors
+    ///
+    /// Will error if `value` is larger than $2^depth$.
+    pub const fn new(depth: u8, value: u64) -> Result<Self, MerkleError> {
+        match 2u64.checked_pow(depth as u32) {
```

Comment:
This check is the main change, almost everything else are changes to match the new return type.

Reply:
It's probably overkill to check the index every time and add a branch computation, as the responsibility to produce valid indexes falls to the tree implementation, and not the depth/index pair itself. Indexes are often raw scalar values, as it is for example in a `usize` when indexing a vector (that might as well contain a value that is out of bounds).

The way I see, `NodeIndex::new` should be as efficient as possible so we pay the computation price only for the algorithms of the trees. We also need to think of what we are covering here. If the user inputs an invalid index for, let's say, a `Tree::get` function, he will just receive a `None`. So we are not really protecting the user or shielding potential panics.

Having that said, I don't think this part should error or be checked. In case we are convinced this is strictly necessary, I'd recommend using a separated method for that, something like `new_checked`.

Reply:
I can see 3 points on your comment:

1. Performance
2. Semantics of the type
3. Naming

### Performance

I'm guessing this amounts to something like 3 instructions, on the creation of the index struct. Is that something that we really have to worry about?

### Semantics

This still behaves like the `usize` used to index on a vector. It just introduces internal consistency to the type, it does not guarantee the value is in-bounds.

### Naming

changed

Reply:
Regarding the performance topic, the 4 instructions aren't problematic, but we introduce branching for every node index instantiation + panic/unwind check. This can lead to big performance drop.

Reply:
@vlopes11 I don't understand what you're tying to say. The 4 instructions that I was talking are the `pow`, the overflow branch, the comparison, and the branch for the comparison. There is no panicking in the code.

I'm not seeing were the performance drop would be, could you clear that out?

Reply:
If you add a `Result`, you force the user to `match` the return. Even if he unwraps, it is still a branch.

This means an operation like `NodeIndex::new(a, b)` will no longer be just 2 integers moved to the stack; it will, instead, be a branch to match its output and maybe unwind if the input is invalid.

Even tho the `new` itself is more expensive, the additional cost is not restricted to that and will leak to the consumers of this function.